### PR TITLE
Avoid linking AO aux tables to extension

### PR DESCRIPTION
**Pull request by InnerLife0**
Avoid linking AO aux tables to extension

Unlike TOAST tables, AO aux tables are [linked](https://github.com/arenadata/gpdb/blob/1710ca20c6de81dd533b3d40e3fe0c1e1544daa1/src/backend/catalog/heap.c#L1802) to extension during its creation.
When table is dropped from extension using ALTER, aux tables are left as
extension dependencies. It means we can't DROP a table after that, because AO
aux tables are still used by extension. In other words, this behavior makes
impossible to DROP AO table that was created by extension without extension
DROP.

The goal of the patch is to change this behavior, allowing to DROP a table.
Unnecessary dependencies are now not created in heap_create_with_catalog(). The
behavior is the same as for TOAST tables, which are aux tables too. The
dependency [created](https://github.com/arenadata/gpdb/blob/1710ca20c6de81dd533b3d40e3fe0c1e1544daa1/src/backend/catalog/aocatalog.c#L235) in CreateAOAuxiliaryTable() is enough.

The patch contains a test that describes target behavior. However, the test can
be run only manually. GPDB is not ready to run tests in modules subdir
automatically.

___________
The impossibility of running tests automatically is an old [story](https://github.com/greenplum-db/gpdb/pull/14534) and should be resolved in a separate PR.

The example of test running is:
```
REGRESS_OPTS=--init-file=$(pwd)/src/test/regress/init_file make -C src/test/modules/test_extensions installcheck
```
The attentive reviewer may find that I don't test blockdir. This is done intentionally to minimize patch diff. There are no test extensions with index and I wanted to avoid creation of our own. I hope this will not be a problem during review as changes are simple and intuitive.

### CODE REVIEW

**Comment by red1452** [comment]
Why do you remove all dependencies for AO tables, as it was done for toast tables? Maybe, is it better to move your check to the call of function `recordDependencyOnCurrentExtension` (do not execute it if table is AO) and do not touch other dependencies?

File: src/backend/catalog/heap.c [hunk 1]
```diff
@@ -1778,9 +1778,14 @@ heap_create_with_catalog(const char *relname,
...
 	 * Also, skip this in bootstrap mode, since we don't make dependencies
 	 * while bootstrapping.
+	 *
+	 * GPDB: The section about TOAST is still relevant for AO aux tables.
 	 */
 	if (relkind != RELKIND_COMPOSITE_TYPE &&
 		relkind != RELKIND_TOASTVALUE &&
+		relkind != RELKIND_AOSEGMENTS &&
+		relkind != RELKIND_AOBLOCKDIR &&
+		relkind != RELKIND_AOVISIMAP &&
```
Comment (red1452) [inline src/backend/catalog/heap.c]: You may use `IsAppendonlyMetadataRelkind` function. It checks the same `relkind`.

**Comment by RekGRpth** [comment]
> Unnecessary dependencies are now not created in heap_create_with_catalog().

Did I understand correctly, and this is the expected behavior: if an extension with an AO table was created BEFORE the patch, then we still will not be able to correctly remove the table from the extension even AFTER applying the patch?

**Comment by InnerLife0** [comment]
> Why do you remove all dependencies for AO tables, as it was done for toast tables? Maybe, is it better to move your check to the call of function `recordDependencyOnCurrentExtension` (do not execute it if table is AO) and do not touch other dependencies?

The idea of patched block is to add some _extra_ dependencies to rels with some kinds. As described, some kinds doesn't need such dependencies.
Here I will try to extend each point of in-code comment.
1. AO aux tables have own namespace. I have nothing to add here. We don't need extra dependency as tables already in own isolated namespace.
2. AO aux tables are depend on parent tables. Thus, they doesn't need their own dependency to owner. Parent is already linked to owner. So, aux table is linked to owner through parent. This works the same for TOAST.
3. AO aux tables have no built-in acl to create it during table creation. The idea of creating ACL to aux table is questionable by its design, but anyway, it's true we can use some after table creation. But we have nothing to do with this during table creation by the design. To prove this, you may look how `relacl` is used above. It's always NULL for aux tables.

As a conclusion, we have a block of code, that do excessive job (or even do nothing). This code adds some extra dependencies, which are not related to aux tables (doesn't matter, AO or TOAST). We want to avoid this job for AO aux tables like it's done for TOAST. This mentioned in code and in commit description. As our goal is to change extension-related behavior (and this is what actually changed) the commit is named like so.

> Did I understand correctly, and this is the expected behavior: if an extension with an AO table was created BEFORE the patch, then we still will not be able to correctly remove the table from the extension even AFTER applying the patch?

You understand this right and this is the expected behavior. Current patch has nothing to do with already installed extensions. Our main goal is to cover our own back for future extensions with AO tables.

File: src/backend/catalog/heap.c (hunk 1 above)
Comment (InnerLife0) [reply src/backend/catalog/heap.c]: Done. Thanks.

**Review by RekGRpth — APPROVED** [review]

**Review by red1452 — APPROVED** [review]

**Pull request merged**